File: Reverse_Linked_List.cc
/*
Reverse a singly linked list.

click to show more hints.

Hint:
A linked list can be reversed either iteratively or recursively. Could you implement both?
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution 
{
public:
    ListNode* reverseList(ListNode* head) 
    {
        if (head == NULL || head->next == NULL)
            return head;
        
        ListNode* pPre = NULL;
        ListNode* pCur = head;
        ListNode* pNext = pCur->next;
        
        while (pCur != NULL)
        {
            pNext = pCur->next; 
            pCur->next = pPre; 
            pPre = pCur;
            pCur = pNext;
        }
        head = pPre;
        
        return head;
    }
};

File: go/reverse-linked-list/reverse-linked-list.go
package reverse_linked_list

import "fmt"

/*
Reverse a singly linked list.

click to show more hints.

Hint:
A linked list can be reversed either iteratively or recursively.
Could you implement both?
*/

type ListNode struct {
	Val  int
	Next *ListNode
}

// 头插法，就倒置了
func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	dummy := &ListNode{-1, head}
	cur := head.Next // 第二个结点，从第二个结点开始，头插法
	head.Next = nil  // 倒置后的尾结点

	for cur != nil {
		tmp := cur.Next

		cur.Next = dummy.Next
		dummy.Next = cur

		cur = tmp
	}

	return dummy.Next
}
